File: Sharebook/Server/src/Services/Identity/Sharebook.Identity.API/Client/Validators/AccountSettingsClientParametersValidator.cs
namespace Sharebook.Identity.API.Client.Validators
{
    using System;

    using Sharebook.Common.Extensions;
    using Sharebook.Common.Settings;
    using Sharebook.Identity.API.Application.Parameters;

    public interface IAccountSettingsClientParametersValidator
    {
        void Validate(IGetAccountSettingsParameters parameters);

        void Validate(IUpdateAccountSettingsParameters parameters);

        void Validate(IUpdateProfileParameters parameters);

        void Validate(IUpdateProfileImageParameters parameters);

        void Validate(IUpdateProfileBannerParameters parameters);

        void Validate(IRemoveProfileBannerParameters parameters);
    }

    public class AccountSettingsClientParametersValidator : IAccountSettingsClientParametersValidator
    {
        private readonly IAccountSettingsClientRequiredParametersValidator accountSettingsClientRequiredParametersValidator;
        private readonly ITwitterClient client;

        public AccountSettingsClientParametersValidator(ITwitterClient client, IAccountSettingsClientRequiredParametersValidator accountSettingsClientRequiredParametersValidator)
        {
            this.client = client;
            this.accountSettingsClientRequiredParametersValidator = accountSettingsClientRequiredParametersValidator;
        }

        private WorldFeedLimits Limits => this.client.Config.Limits;

        public void Validate(IGetAccountSettingsParameters parameters)
        {
            this.accountSettingsClientRequiredParametersValidator.Validate(parameters);
        }

        public void Validate(IUpdateAccountSettingsParameters parameters)
        {
            this.accountSettingsClientRequiredParametersValidator.Validate(parameters);

            if (!parameters.DisplayLanguage.IsADisplayLanguage())
            {
                throw new ArgumentException("As of 2019-10-06 this language is not supported by Twitter", $"{nameof(parameters.DisplayLanguage)}");
            }
        }

        public void Validate(IUpdateProfileParameters parameters)
        {
            this.accountSettingsClientRequiredParametersValidator.Validate(parameters);

            ThrowIfParameterSizeIsInvalid(parameters.Name, $"{nameof(parameters.Name)}", Limits.ACCOUNT_SETTINGS_PROFILE_NAME_MAX_LENGTH);
            ThrowIfParameterSizeIsInvalid(parameters.Description, $"{nameof(parameters.Description)}", Limits.ACCOUNT_SETTINGS_PROFILE_DESCRIPTION_MAX_LENGTH);
            ThrowIfParameterSizeIsInvalid(parameters.Location, $"{nameof(parameters.Location)}", Limits.ACCOUNT_SETTINGS_PROFILE_LOCATION_MAX_LENGTH);
            ThrowIfParameterSizeIsInvalid(parameters.WebsiteUrl, $"{nameof(parameters.WebsiteUrl)}", Limits.ACCOUNT_SETTINGS_PROFILE_WEBSITE_URL_MAX_LENGTH);
        }

        public void Validate(IUpdateProfileImageParameters parameters)
        {
            this.accountSettingsClientRequiredParametersValidator.Validate(parameters);
        }

        public void Validate(IUpdateProfileBannerParameters parameters)
        {
            this.accountSettingsClientRequiredParametersValidator.Validate(parameters);
        }

        public void Validate(IRemoveProfileBannerParameters parameters)
        {
            this.accountSettingsClientRequiredParametersValidator.Validate(parameters);
        }

        // ReSharper disable once ParameterOnlyUsedForPreconditionCheck.Local
        private static void ThrowIfParameterSizeIsInvalid(string value, string parameterName, int maxSize)
        {
            if (value != null && value.Length > maxSize)
            {
                throw new ArgumentException($"{parameterName} cannot contain more than {maxSize} characters.", parameterName);
            }
        }
    }
}

File: Sharebook/Client/src/app/core/Core/Client/Validators/AccountSettingsClientParametersValidator.ts
import {Inject, Injectable, InjectionToken} from "@angular/core";

import ArgumentException from "../../../../c#-objects/TypeScript.NET-Core/packages/Core/source/Exceptions/ArgumentException";
import {SharebookLimits} from "../../../Public/Settings/SharebookLimits";
import {IGetAccountSettingsParameters} from "../../../Public/Parameters/AccountSettingsClient/GetAccountSettingsParameters";
import {IUpdateAccountSettingsParameters} from "../../../Public/Parameters/AccountSettingsClient/UpdateAccountSettingsParameters";
import {IUpdateProfileParameters} from "../../../Public/Parameters/AccountSettingsClient/UpdateProfileParameters";
import {IUpdateProfileImageParameters} from "../../../Public/Parameters/AccountSettingsClient/UpdateProfileImageParameters";
import {IUpdateProfileBannerParameters} from "../../../Public/Parameters/AccountSettingsClient/UpdateProfileBannerParameters";
import {IRemoveProfileBannerParameters} from "../../../Public/Parameters/AccountSettingsClient/RemoveProfileBannerParameters";
import {
  AccountSettingsClientRequiredParametersValidator,
  IAccountSettingsClientRequiredParametersValidator,
  IAccountSettingsClientRequiredParametersValidatorToken
} from "./AccountSettingsClientRequiredParametersValidator";
import {ITwitterClient, ITwitterClientToken} from "../../../Public/ITwitterClient";
import {AccountActivityParameters} from "./parameters-types";
import {TwitterClient} from "../../../../sharebook/TwitterClient";

export interface IAccountSettingsClientParametersValidator {
  validate(parameters: IGetAccountSettingsParameters): void;

  validate(parameters: IUpdateAccountSettingsParameters): void;

  validate(parameters: IUpdateProfileParameters): void;

  validate(parameters: IUpdateProfileImageParameters): void;

  validate(parameters: IUpdateProfileBannerParameters): void;

  validate(parameters: IRemoveProfileBannerParameters): void;
}

export const IAccountSettingsClientParametersValidatorToken = new InjectionToken<IAccountSettingsClientParametersValidator>('IAccountSettingsClientParametersValidator', {
  providedIn: 'root',
  factory: () => new AccountSettingsClientParametersValidator(Inject(TwitterClient), Inject(AccountSettingsClientRequiredParametersValidator)),
});

@Injectable()
export class AccountSettingsClientParametersValidator implements IAccountSettingsClientParametersValidator {
  private readonly _accountSettingsClientRequiredParametersValidator: IAccountSettingsClientRequiredParametersValidator;
  private readonly _client: ITwitterClient;

  constructor(@Inject(ITwitterClientToken) client: ITwitterClient,
              @Inject(IAccountSettingsClientRequiredParametersValidatorToken) accountSettingsClientRequiredParametersValidator: IAccountSettingsClientRequiredParametersValidator) {
    this._client = client;
    this._accountSettingsClientRequiredParametersValidator = accountSettingsClientRequiredParametersValidator;
  }

  private get Limits(): SharebookLimits {
    return this._client.config.limits;
  }

  public validate(parameters: AccountActivityParameters): void {
    this._accountSettingsClientRequiredParametersValidator.validate(parameters);

    if (AccountSettingsClientParametersValidator.isIUpdateProfileParameters(parameters)) {
      AccountSettingsClientParametersValidator.throwIfParameterSizeIsInvalid(parameters.name, `${nameof(parameters.name)}`, this.Limits.ACCOUNT_SETTINGS_PROFILE_NAME_MAX_LENGTH);
      AccountSettingsClientParametersValidator.throwIfParameterSizeIsInvalid(parameters.description, `${nameof(parameters.description)}`, this.Limits.ACCOUNT_SETTINGS_PROFILE_DESCRIPTION_MAX_LENGTH);
      AccountSettingsClientParametersValidator.throwIfParameterSizeIsInvalid(parameters.location, `${nameof(parameters.location)}`, this.Limits.ACCOUNT_SETTINGS_PROFILE_LOCATION_MAX_LENGTH);
      AccountSettingsClientParametersValidator.throwIfParameterSizeIsInvalid(parameters.websiteUrl, `${nameof(parameters.websiteUrl)}`, this.Limits.ACCOUNT_SETTINGS_PROFILE_WEBSITE_URL_MAX_LENGTH);
    }
  }

  private static throwIfParameterSizeIsInvalid(value: string, parameterName: string, maxSize: number): void {
    if (value != null && value.length > maxSize) {
      throw new ArgumentException(`${parameterName} cannot contain more than ${maxSize} characters.`, parameterName);
    }
  }

  private static isIUpdateProfileParameters(parameters: AccountActivityParameters): parameters is IUpdateProfileParameters {
    return (parameters as IUpdateProfileParameters).description !== undefined;
  }
}
